sbin/sysctl/sysctl.8
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sysctl.8	8.2 (Berkeley) 5/9/95
.\"
.Dd $Mdocdate: August 26 2008 $
.Dt SYSCTL 8
.Os
.Sh NAME
.Nm sysctl
.Nd get or set kernel state
.Sh SYNOPSIS
.Nm sysctl
.Op Fl Aan
.Nm sysctl
.Op Fl n
.Ar name ...
.Nm sysctl
.Op Fl nq
.Ar name Ns = Ns Ar value ...
.Sh DESCRIPTION
The
.Nm
utility retrieves kernel state and allows processes with
appropriate privilege to set kernel state.
The state to be retrieved or set is described using a
.Dq Management Information Base
.Pq MIB
style name, described as a dotted set of components.
.Pp
When retrieving a variable,
a subset of the MIB name may be specified to retrieve a list of
variables in that subset.
For example, to list all the machdep variables:
.Pp
.Dl $ sysctl machdep
.Pp
When setting a variable,
the MIB name should be followed by an equal sign and the new value.
.Pp
The options are as follows:
.Bl -tag -width xxx
.It Fl A
List all the known MIB names including tables.
Those with string or integer values will be printed as with the
.Fl a
flag; for the table values, the name of the utility to retrieve them is given.
.It Fl a
List all the currently available string or integer values.
This is the default, if no parameters are given to
.Nm .
.It Fl n
Suppress printing of the field name, only output the field value.
Useful for setting shell variables.
For example, to set the psize shell variable to the pagesize of the hardware:
.Pp
.Dl # set psize=`sysctl -n hw.pagesize`
.Pp
.It Fl q
Suppress all output when setting a variable.
This option overrides the behaviour of
.Fl n .
.It Ar name Ns = Ns Ar value
Attempt to set the specified variable
.Ar name
to
.Ar value .
.El
.Pp
The information available from
.Nm
consists of integers, strings, and tables.
The tabular information can only be retrieved by special
purpose programs such as
.Xr ps 1 ,
.Xr systat 1 ,
and
.Xr netstat 1 .
The string and integer information is summarized below.
For a detailed description of these variables, see
.Xr sysctl 3 .
The changeable column indicates whether a process with appropriate
privilege can change the value.
.Pp
Note:
not all of the variables are relevant to all architectures.
.Bl -column net.inet.ip.ipsec-expire-acquirexxxx integerxxx
.It Sy Name	Type	Changeable
.It kern.ostype	string	no
.It kern.osrelease	string	no
.It kern.osrevision	integer	no
.It kern.version	string	no
.It kern.maxvnodes	integer	yes
.It kern.maxproc	integer	yes
.It kern.maxfiles	integer	yes
.It kern.argmax	integer	no
.It kern.securelevel	integer	raise only
.It kern.hostname	string	yes
.It kern.hostid	u_int	yes
.It kern.clockrate	struct	no
.It kern.posix1version	integer	no
.It kern.ngroups	integer	no
.It kern.job_control	integer	no
.It kern.saved_ids	integer	no
.It kern.boottime	struct	no
.It kern.domainname	string	yes
.It kern.maxpartitions	integer	no
.It kern.rawpartition	integer	no
.It kern.osversion	string	no
.It kern.somaxconn	integer	yes
.It kern.sominconn	integer	yes
.It kern.usermount	integer	yes
.It kern.random	struct	no
.It kern.nosuidcoredump	integer	yes
.It kern.fsync	integer	no
.It kern.sysvmsg	integer	no
.It kern.sysvsem	integer	no
.It kern.sysvshm	integer	no
.It kern.arandom	u_int	no
.It kern.msgbufsize	integer	no
.It kern.malloc.buckets	string	no
.It kern.malloc.bucket.<sz>	string	no
.It kern.malloc.kmemnames	string	no
.It kern.malloc.kmemstat.<name>	string	no
.It kern.cp_time	struct	no
.It kern.nchstats	struct	no
.It kern.forkstat	struct	no
.It kern.nselcoll	integer	no
.It kern.tty.tk_nin	int64_t	no
.It kern.tty.tk_nout	int64_t	no
.It kern.tty.tk_rawcc	int64_t	no
.It kern.tty.tk_cancc	int64_t	no
.It kern.tty.ttyinfo	struct	no
.It kern.tty.maxptys	integer	yes
.It kern.tty.nptys	integer	no
.It kern.ccpu	u_int	no
.It kern.fscale	integer	no
.It kern.nprocs	integer	no
.It kern.stackgap_random	integer	yes
.It kern.usercrypto	integer	yes
.It kern.cryptodevallowsoft	integer	yes
.It kern.splassert	integer	yes
.It kern.nfiles	integer	no
.It kern.ttycount	integer	no
.It kern.numvnodes	integer	no
.It kern.userasymcrypto	integer	yes
.It kern.seminfo.semmni	integer	yes
.It kern.seminfo.semmns	integer	yes
.It kern.seminfo.semmnu	integer	yes
.It kern.seminfo.semmsl	integer	yes
.It kern.seminfo.semopm	integer	yes
.It kern.seminfo.semume	integer	no
.It kern.seminfo.semusz	integer	no
.It kern.seminfo.semvmx	integer	no
.It kern.seminfo.semaem	integer	no
.It kern.shminfo.shmmax	integer	yes
.It kern.shminfo.shmmin	integer	yes
.It kern.shminfo.shmmni	integer	yes
.It kern.shminfo.shmseg	integer	yes
.It kern.shminfo.shmall	integer	yes
.It kern.watchdog.period	integer	yes
.It kern.watchdog.auto	integer	yes
.It kern.emul.nemuls	integer	no
.It kern.emul.other	integer	yes
.It kern.maxclusters	integer	yes
.It kern.timecounter.tick	integer	no
.It kern.timecounter.timestepwarnings	integer	yes
.It kern.timecounter.hardware	string	yes
.It kern.timecounter.choice	string	no
.It kern.rootdev	integer	no
.It kern.swapdev	integer	no
.It kern.dumpdev	integer	no
.It kern.dumplo	quad	no
.It kern.dumpmag	integer	no
.It vm.vmmeter	struct	no
.It vm.loadavg	struct	no
.It vm.psstrings	struct	no
.It vm.uvmexp	struct	no
.It vm.swapencrypt.enable	integer	yes
.It vm.swapencrypt.keyscreated	integer	no
.It vm.swapencrypt.keysdeleted	integer	no
.It vm.nkmempages	integer	no
.It vm.anonmin	integer	yes
.It vm.vtextmin	integer	yes
.It vm.vnodemin	integer	yes
.It vm.maxslp	integer	no
.It vm.uspace	integer	no
.It fs.posix.setuid	integer	yes
.It net.inet.ip.forwarding	integer	yes
.It net.inet.ip.redirect	integer	yes
.It net.inet.ip.ttl	integer	yes
.\" .It net.inet.ip.mtu	integer	yes
.It net.inet.ip.sourceroute	integer	yes
.It net.inet.ip.directed-broadcast	integer	yes
.It net.inet.ip.portfirst	integer	yes
.It net.inet.ip.portlast	integer	yes
.It net.inet.ip.porthifirst	integer	yes
.It net.inet.ip.porthilast	integer	yes
.It net.inet.ip.maxqueue	integer	yes
.It net.inet.ip.encdebug	integer	yes
.It net.inet.ip.ipsec-expire-acquire	integer	yes
.It net.inet.ip.ipsec-invalid-life	integer	yes
.It net.inet.ip.ipsec-pfs	integer	yes
.It net.inet.ip.ipsec-soft-allocs	integer	yes
.It net.inet.ip.ipsec-allocs	integer	yes
.It net.inet.ip.ipsec-soft-bytes	integer	yes
.It net.inet.ip.ipsec-bytes	integer	yes
.It net.inet.ip.ipsec-timeout	integer	yes
.It net.inet.ip.ipsec-soft-timeout	integer	yes
.It net.inet.ip.ipsec-soft-firstuse	integer	yes
.It net.inet.ip.ipsec-firstuse	integer	yes
.It net.inet.ip.ipsec-enc-alg	string	yes
.It net.inet.ip.ipsec-auth-alg	string	yes
.It net.inet.ip.mtudisc	integer	yes
.It net.inet.ip.mtudisctimeout	integer	yes
.It net.inet.ip.ipsec-comp-alg	string	yes
.It net.inet.ip.mforwarding	integer	yes
.It net.inet.ip.multipath	integer	yes
.It net.inet.icmp.maskrepl	integer	yes
.It net.inet.icmp.bmcastecho	integer	yes
.It net.inet.icmp.errppslimit	integer	yes
.It net.inet.icmp.rediraccept	integer	yes
.It net.inet.icmp.redirtimeout	integer	yes
.It net.inet.icmp.tstamprepl	integer	yes
.It net.inet.ipip.allow	integer	yes
.It net.inet.tcp.rfc1323	integer	yes
.It net.inet.tcp.keepinittime	integer	yes
.It net.inet.tcp.keepidle	integer	yes
.It net.inet.tcp.keepintvl	integer	yes
.It net.inet.tcp.slowhz	integer	no
.It net.inet.tcp.baddynamic	array	yes
.It net.inet.tcp.recvspace	integer	yes
.It net.inet.tcp.sendspace	integer	yes
.It net.inet.tcp.sack	integer	yes
.It net.inet.tcp.mssdflt	integer	yes
.It net.inet.tcp.rstppslimit	integer	yes
.It net.inet.tcp.ackonpush	integer	yes
.It net.inet.tcp.ecn	integer	yes
.It net.inet.tcp.syncachelimit	integer	yes
.It net.inet.tcp.synbucketlimit	integer	yes
.It net.inet.tcp.rfc3390	integer	yes
.It net.inet.tcp.reasslimit	integer	yes
.It net.inet.udp.checksum	integer	yes
.It net.inet.udp.baddynamic	array	yes
.It net.inet.udp.recvspace	integer	yes
.It net.inet.udp.sendspace	integer	yes
.It net.inet.gre.allow	integer	yes
.It net.inet.gre.wccp	integer	yes
.It net.inet.esp.enable	integer	yes
.It net.inet.esp.udpencap	integer	yes
.It net.inet.esp.udpencap_port	integer	yes
.It net.inet.ah.enable	integer	yes
.It net.inet.mobileip.allow	integer	yes
.It net.inet.etherip.allow	integer	yes
.It net.inet.ipcomp.enable	integer	yes
.It net.inet.carp.allow	integer	yes
.It net.inet.carp.preempt	integer	yes
.It net.inet.carp.log	integer	yes
.It net.inet6.ip6.forwarding	integer	yes
.It net.inet6.ip6.redirect	integer	yes
.It net.inet6.ip6.hlim	integer	yes
.It net.inet6.ip6.maxfragpackets	integer	yes
.It net.inet6.ip6.accept_rtadv	integer	yes
.It net.inet6.ip6.keepfaith	integer	yes
.It net.inet6.ip6.log_interval	integer	yes
.It net.inet6.ip6.hdrnestlimit	integer	yes
.It net.inet6.ip6.dad_count	integer	yes
.It net.inet6.ip6.auto_flowlabel	integer	yes
.It net.inet6.ip6.defmcasthlim	integer	yes
.It net.inet6.ip6.kame_version	string	no
.It net.inet6.ip6.use_deprecated	integer	yes
.It net.inet6.ip6.rr_prune	integer	yes
.It net.inet6.ip6.v6only	integer	no
.It net.inet6.ip6.maxfrags	integer	yes
.It net.inet6.ip6.mforwarding	integer	yes

usr.bin/kdump/sysctl.c
/*
 * Copyright (c) 2006 Michael Shalayeff
 * All rights reserved.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF MIND, USE, DATA OR PROFITS, WHETHER IN
 * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
 * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef lint
static const char rcsid[] = "$ABSD: sysctl.c,v 1.2 2010/06/05 11:33:16 mickey Exp $";
#endif /* not lint */

#include <sys/param.h>
#include <sys/sysctl.h>
#include <sys/socket.h>
#include <sys/malloc.h>
#include <sys/uio.h>
#include <sys/namei.h>
#include <sys/vmmeter.h>
#include <sys/tty.h>
#include <sys/sem.h>
#include <sys/shm.h>
#include <sys/mount.h>

#include <net/if.h>
#include <netinet/in.h>
#include <netinet/in_systm.h>
#include <netinet/ip.h>
#include <netinet/ip_var.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp_var.h>
#include <netinet/tcp.h>
#include <netinet/tcp_timer.h>
#include <netinet/tcp_var.h>
#include <netinet/udp.h>
#include <netinet/udp_var.h>
#include <netinet/ip_ipip.h>
#include <netinet/ip_gre.h>
#include <netinet/ip_esp.h>
#include <netinet/ip_ah.h>
#include <netinet/ip_ipcomp.h>
#include <netinet/ip_ether.h>
#include <netinet/ip_carp.h>
#include <netinet/icmp6.h>
#include <netinet6/pim6_var.h>

#include <ddb/db_var.h>

#include <uvm/uvm_swap_encrypt.h>

#include <ufs/ufs/dinode.h>
#include <ufs/ffs/ffs_extern.h>

#include <nfs/rpcv2.h>
#include <nfs/nfsproto.h>
#include <nfs/nfs.h>

#include <string.h>
#include <stdio.h>

#include <machine/cpu.h>
#ifndef	CTL_MACHDEP_NAMES
#define	CTL_MACHDEP_NAMES	{ 0 }
#endif
#ifndef	CPU_MAXID
#define	CPU_MAXID		0
#endif

#define	EMUL_MAXID		16

const struct ctlname root[] = CTL_NAMES;
const struct ctlname kern[] = CTL_KERN_NAMES;
const struct ctlname kmalloc[] = CTL_KERN_MALLOC_NAMES;
      char *kmemnames[] = INITKMEMNAMES;
      struct ctlname kmems[M_LAST];
const struct ctlname nchstats[] = CTL_KERN_NCHSTATS_NAMES;
const struct ctlname forkstat[] = CTL_KERN_FORKSTAT_NAMES;
const struct ctlname tty[] = CTL_KERN_TTY_NAMES;
const struct ctlname seminfo[] = CTL_KERN_SEMINFO_NAMES;
const struct ctlname shminfo[] = CTL_KERN_SHMINFO_NAMES;
const struct ctlname intrcnt[] = CTL_KERN_INTRCNT_NAMES;
const struct ctlname wdog[] = CTL_KERN_WATCHDOG_NAMES;
const struct ctlname tcount[] = CTL_KERN_TIMECOUNTER_NAMES;
      struct ctlname emul[EMUL_MAXID] = {{ "nemuls",	0 }};
const struct ctlname emulnames[] = CTL_KERN_EMUL_NAMES;
const struct ctlname procargs[] = CTL_KERN_PROCARGS_NAMES;
const struct ctlname vm[] = CTL_VM_NAMES;
const struct ctlname vmswen[] = CTL_SWPENC_NAMES;
const struct ctlname fs[] = CTL_FS_NAMES;
const struct ctlname posix[] = CTL_FS_POSIX_NAMES;
const struct ctlname net[] = CTL_NET_NAMES;
const struct ctlname ipproto[] = CTL_IPPROTO_NAMES;
const struct ctlname ip[] = IPCTL_NAMES;
const struct ctlname ifq[] = CTL_IFQ_NAMES;
const struct ctlname icmp[] = ICMPCTL_NAMES;
const struct ctlname ipip[] = IPIPCTL_NAMES;
const struct ctlname tcp[] = TCPCTL_NAMES;
const struct ctlname udp[] = UDPCTL_NAMES;
const struct ctlname gre[] = GRECTL_NAMES;
const struct ctlname esp[] = ESPCTL_NAMES;
const struct ctlname ah[] = AHCTL_NAMES;
const struct ctlname mip[] = MOBILEIPCTL_NAMES;
const struct ctlname eip[] = ETHERIPCTL_NAMES;
const struct ctlname ipcomp[] = IPCOMPCTL_NAMES;
const struct ctlname carp[] = CARPCTL_NAMES;
const struct ctlname route[] = CTL_NET_RT_NAMES;
const struct ctlname inet6[] = IPV6CTL_NAMES;
const struct ctlname ip6proto[] = CTL_IPV6PROTO_NAMES;
const struct ctlname ip6[] = IPV6CTL_NAMES;
const struct ctlname icmp6[] = ICMPV6CTL_NAMES;
const struct ctlname pim6[] = PIM6CTL_NAMES;
const struct ctlname pfkey[] = CTL_NET_KEY_NAMES;
const struct ctlname bpf[] = CTL_NET_BPF_NAMES;
const struct ctlname debug[] =  { 0 };
const struct ctlname hw[] = CTL_HW_NAMES;
const struct ctlname md[] = CTL_MACHDEP_NAMES;
const struct ctlname user[] = CTL_USER_NAMES;
const struct ctlname ddb[] = CTL_DDB_NAMES;
const struct ctlname vfs[] = {
	{ "gen",	0 },	/* fake list to just produce the names */
	{ "ffs",	0 },
	{ "nfs",	0 },
	{ "mfs",	0 },
	{ "msdos",	0 },
	{ "lfs",	0 },
	{ "ntfs",	0 },
	{ "fdesc",	0 },
	{ "portal",	0 },
	{ "null",	0 },
	{ "umap",	0 },
	{ "kern",	0 },
	{ "proc",	0 },
	{ "udf",	0 },
	{ "cd9660",	0 },
	{ "union",	0 },
	{ "ados",	0 },
	{ "ext2",	0 },
	{ NULL },	/* 18 */
	{ NULL },	/* 19 */
	{ NULL },	/* 20 */
	{ "xfs",	0 },
};
const struct ctlname genfs[] = CTL_VFSGENCTL_NAMES;
const struct ctlname ffsname[] = FFS_NAMES;
const struct ctlname nfsname[] = FS_NFS_NAMES;
const struct ctlname empty[] = {{ NULL }};

struct ctlist {
	const struct ctlname *names;
	const struct ctlist *lists;
	int maxid;
};

const struct ctlist kmemlists[] = {
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ kmems,	NULL,	M_LAST }
}, emulists[EMUL_MAXID] = {
	{ NULL },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
	{ emulnames,	NULL,	KERN_EMUL_MAXID },
}, pargslist[] = {
	{ procargs,	NULL,	KERN_PROCARGS_MAXID }
}, kernlists[KERN_MAXID] = {
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* proc */
	{ NULL },
	{ NULL },	/* prof */
	{ NULL },	/* posix */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ kmalloc,	kmemlists,	KERN_MALLOC_MAXID },
	{ NULL },
	{ nchstats,	NULL,	KERN_NCHSTATS_MAXID },
	{ forkstat,	NULL,	KERN_FORKSTAT_MAXID },
	{ NULL },
	{ tty,		NULL,	KERN_TTY_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* pool */
	{ NULL },
	{ NULL },	/* ipc */
	{ NULL },
	{ NULL },
	{ NULL },
	{ empty,	pargslist,	-1 },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ seminfo,	NULL,	KERN_SEMINFO_MAXID },
	{ shminfo,	NULL,	KERN_SHMINFO_MAXID },
	{ intrcnt,	NULL,	KERN_INTRCNT_MAXID  },
	{ wdog,		NULL,	KERN_WATCHDOG_MAXID },
	{ emul,		emulists,	EMUL_MAXID },
	{ NULL },	/* proc2 */
	{ NULL },
	{ NULL },	/* evcount */
	{ tcount,	NULL,	KERN_TIMECOUNTER_MAXID },
}, vmlists[VM_MAXID] = {
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ vmswen,	NULL,	SWPENC_MAXID },
}, fslists[FS_MAXID] = {
	{ NULL },
	{ posix,	NULL,	FS_POSIX_MAXID },
}, iplists[IPCTL_MAXID] = {
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ ifq,		NULL,	IFQCTL_MAXID },
	{ NULL },
	{ NULL },
}, inetlists[IPPROTO_MAXID] = {
	{ ip,		iplists,IPPROTO_MAXID },
	{ icmp,		NULL,	ICMPCTL_MAXID },
	{ NULL },	/* igmp */
	{ NULL },	/* ggp */
	{ ipip,		NULL,	IPIPCTL_MAXID },
	{ NULL },
	{ tcp,		NULL,	TCPCTL_MAXID },
	{ NULL },
	{ NULL },	/* egp */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* pup */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ udp,		NULL,	UDPCTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ gre,		NULL,	GRECTL_MAXID },
	{ NULL },
	{ NULL },
	{ esp,		NULL,	ESPCTL_MAXID },
	{ ah,		NULL,	AHCTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ mip,		NULL,	MOBILEIPCTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ eip,		NULL,	ETHERIPCTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ ipcomp,	NULL,	IPCOMPCTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ carp,		NULL,	CARPCTL_MAXID },
}, inet6lists[IPV6PROTO_MAXID] = {
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* tcp6,		NULL,	IPV6PROTO_MAXID }, */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* udp6,		NULL,	IPV6PROTO_MAXID }, */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ ip6,		NULL,	IPV6CTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },	/* ipsec6,	NULL,	IPV6CTL_MAXID }, */
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ icmp6,	NULL,	ICMPV6CTL_MAXID },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ NULL },
	{ pim6,		NULL,	PIM6CTL_MAXID },
}, netlists[NET_MAXID] = {
	{ NULL },
	{ NULL },	/* unix */
	{ ipproto,	inetlists,	IPPROTO_MAXID },
	{ NULL },	/* implink */
	{ NULL },	/* pup */
	{ NULL },	/* chaos */
	{ NULL },	/* xns */
	{ NULL },	/* iso */
	{ NULL },	/* ymca */
	{ NULL },	/* datakit */
	{ NULL },	/* ccitt */
	{ NULL },	/* sna */
	{ NULL },	/* decnet */
	{ NULL },	/* dli */
	{ NULL },	/* lat */
	{ NULL },	/* hylink */
	{ NULL },	/* appletalk */
	{ route,	NULL,	NET_RT_MAXID },
	{ NULL },	/* link */
	{ NULL },	/* xtp */
	{ NULL },	/* coip */
	{ NULL },	/* cnt */
	{ NULL },	/* rtip */
	{ NULL },	/* ipx */
	{ ip6proto,	inet6lists,	IPV6PROTO_MAXID },
	{ NULL },	/* pip */
	{ NULL },	/* isdn */
	{ NULL },	/* natm */
	{ NULL },	/* encap */
	{ NULL },	/* sip */
	{ pfkey,	NULL,	NET_KEY_MAXID },
	{ bpf,		NULL,	NET_BPF_MAXID }
}, vfslists[] = {
	/* this order comes fomr vfsconf.vfc_typenum */
	{ genfs,	NULL,	VFS_MAXID },
	{ ffsname,	NULL,	FFS_MAXID },
	{ nfsname,	NULL,	NFS_MAXID },
	{ NULL },	/* mfs */
	{ NULL },	/* msdos */
	{ NULL },	/* lfs */
	{ NULL },	/* ntfs */
	{ NULL },	/* fdesc */
	{ NULL },	/* portal */
	{ NULL },	/* null */
	{ NULL },	/* umap */
	{ NULL },	/* kern */
	{ NULL },	/* proc */
	{ NULL },	/* udf */
	{ NULL },	/* cd9660 */
	{ NULL },	/* union */
	{ NULL },	/* ados */
	{ NULL },	/* ext2 */
	{ NULL },	/* 18 */
	{ NULL },	/* 19 */
	{ NULL },	/* 20 */
	{ NULL },	/* xfs */
};

const struct ctlist rootlists[] = {
	{ NULL },
	{ kern,	kernlists,	KERN_MAXID },
	{ vm,	vmlists,	VM_MAXID },
	{ fs,	fslists,	FS_MAXID },
	{ net,	netlists,	NET_MAXID },
	{ debug,NULL,		0 },
	{ hw,	NULL,		HW_MAXID },
	{ md,	NULL,		CPU_MAXID },
	{ user,	NULL,		USER_MAXID },
	{ ddb,	NULL,		DBCTL_MAXID },
	{ vfs,	vfslists,	sizeof vfslists / sizeof vfslists[0] }
};
struct ctlist rootlist = { root, rootlists, CTL_MAXID };

const char *
sysctlname(int *name, u_int namelen, char *buf, size_t buflen)
{
	const struct ctlist *l;
	const char *p;
	char nbuf[16], *q;
	int i;

	if (kmems[0].ctl_name == NULL)
		for (i = 0; i < M_LAST; i++) {
			if (kmemnames[i] &&
			    (q = kmems[i].ctl_name = strdup(kmemnames[i])))
				for (; *q; q++)
					if (*q == ' ')
						*q = '_';
		}

	if (emul[0].ctl_type == 0) {
		int mib[4];
		size_t len;

		mib[0] = CTL_KERN;
		mib[1] = KERN_EMUL;
		mib[2] = KERN_EMUL_NUM;
		len = sizeof(int);
		if (sysctl(mib, 3, &i, &len, NULL, 0) < 0)
			emul[0].ctl_type = -1;
		else {
			while (i--) {
				mib[2] = i + 1;
				mib[3] = KERN_EMUL_NAME;
				len = sizeof(nbuf);
				if (sysctl(mib, 4, nbuf, &len, NULL, 0) < 0)
					continue;
				emul[i+1].ctl_name = strdup(nbuf);
			}
		}
	}

	buf[0] = '\0';
	for (l = &rootlist, i = 0; i < namelen; name++, i++) {

		if (i)
			strlcat(buf, ".", buflen);

		if (!l || !l->names || *name >= l->maxid ||
		    !l->names[*name].ctl_name) {
			snprintf(nbuf, sizeof nbuf, "%d", *name);
			p = nbuf;
		} else
			p = l->names[*name].ctl_name;

		if (l && l->lists)
			l = &l->lists[l->maxid == -1? 0 : *name];
		else
			l = NULL;

		strlcat(buf, p, buflen);
	}

	return buf;
}
